DPP EasyConnect: Implement required endianness conversions

apahl-cl opened the pull request:
The EasyConnect and EasyMesh specs require network byte order in some places and little endian in other places. This PR adds a bunch of byte order conversions throughout the EasyConnect code to improve spec conformity. 

One area that required a significant refactor was DPP attributes: ID and length fields are expected to be little endian. I introduced a new type, `ec_host_attribute_t`, to describe a DPP attribute where these fields are in host byte order. This will (hopefully) eliminate ambiguity about the byte ordering of a given instance. 

This code depends on the presence of `__BYTE_ORDER` and `__LITTLE_ENDIAN` as predefined macros, and it requires being able to import `byteswap.h`. All of these conditions are satisfied by glibc.

apahl-cl commented:
@bcarlson-dev @tuckerpo-cl

bcarlson-dev commented on inc/ec_base.h:
```diff
@@ -325,6 +354,68 @@ typedef struct {
+ * @param errMsg Format string for error message
+ * @param ... Additional arguments for the format string
+ */
+#define ASSERT_OPT_HAS_VALUE_FREE3(x, ret, ptr1, ptr2, ptr3, errMsg, ...) \
```
I don't think these are needed. If you replace the `if(x == NULL)` with `if(!x)` in the original one, it should still do the same function

bcarlson-dev replied:
If you want the names for clarity, then it might be better to make that change and have them just be defined based on the `ASSERT_NOT_NULL_FREE*` macros

bcarlson-dev commented on tests/test_util.cpp:
```diff
@@ -0,0 +1,66 @@
+#include "util.h" 
+#include <netinet/in.h>
+
+TEST(EmUtilTest, TestStringSplit) {
```
Can you add a few comments to these tests to clarify for those reading it

bcarlson-dev commented on inc/util.h:
```diff
@@ -138,6 +152,26 @@ std::pair<uint8_t, uint8_t> em_freq_to_chan(unsigned int frequency, const std::s
+ * 
+ * @note `ptr` is `void*` because this is typically invoked on data that must be reinterpreted from its assigned type.
+ */
+uint16_t deref_net_uint16_to_host(const void* const ptr);
```
Can you make reference to `htons` and `ntohs` in the documentation for these functions just so people can easily look and be like "okay that is what it is doing"

apahl-cl replied:
Let me know if you had more in mind than what I added here. I basically just wrote enough to be coherent and greppable.

apahl-cl commented:
Pushed a bunch of changes based on your feedback @bcarlson-dev:
- Renamed `ec_attribute_t` to `ec_net_attribute_t`
- Renamed `ec_host_attribute_t` to `ec_attribute_t`
- Added detailed Doxygen comments for each `*_attribute_t`
- Replaced a bunch of `std::optional<...>` with `auto`
- Fixed the use of`free(unwrapped_attrs)` in a few places
- Added a `data` element to the new `ec_attribute_t` as a shorthand for `ec_attribute_t.original->data`
- Added unit testing for `ec_attribute_t.original`
- Added explanatory comments to the `EmUtilByteOrderTest` unit tests

apahl-cl commented:
I made the `ec_attribute_t.original` unit testing a separate commit so it's a bit easier to see in isolation.

bcarlson-dev commented on inc/ec_base.h:
```diff
@@ -228,11 +228,74 @@ typedef struct {
 } ec_credential_object_t;
 
+/**
+ * @brief A DPP attribute as defined in EC 8.1. Can be sent/received over the network.
```
In header files when referencing specifications, can you spell out the spec (EasyConnect) instead of abreviating (EC)
```suggestion
 * @brief A DPP attribute as defined in EasyConnect 8.1. Can be sent/received over the network.
```

bcarlson-dev commented on inc/ec_base.h:
```diff
@@ -228,11 +228,74 @@ typedef struct {
+     */
+    uint16_t length;
+    /**
+     * @brief Points to the `ec_attribute_t` instance this `ec_attribute_t` instance was derived from.
```
Missed in the renaming
```suggestion
     * @brief Points to the `ec_net_attribute_t` instance this `ec_attribute_t` instance was derived from.
```

bcarlson-dev commented on inc/ec_util.h:
```diff
@@ -519,20 +518,20 @@ class ec_util {
+	 * @note Forwards to `unwrap_wrapped_attrib(ec_attribute_t wrapped_attrib, uint8_t *frame, size_t frame_len, uint8_t *frame_attribs, bool uses_aad, uint8_t *key);`
 	 */
-	static std::pair<uint8_t*, uint16_t> unwrap_wrapped_attrib(ec_attribute_t* wrapped_attrib, ec_frame_t *frame, bool uses_aad, uint8_t* key);
+	static std::pair<uint8_t*, uint16_t> unwrap_wrapped_attrib(ec_attribute_t wrapped_attrib, ec_frame_t *frame, bool uses_aad, uint8_t* key);
```
Performance improvement, can you pass by `const` reference here since we are no longer using pointers.
```suggestion
	static std::pair<uint8_t*, uint16_t> unwrap_wrapped_attrib(const ec_attribute_t& wrapped_attrib, ec_frame_t *frame, bool uses_aad, uint8_t* key);
```

bcarlson-dev commented on inc/ec_util.h:
```diff
@@ -551,7 +550,7 @@ class ec_util {
 	 * @note Ensure that the key is valid and the frame is correctly set up before calling this function.
 	 */
-	static std::pair<uint8_t*, uint16_t> unwrap_wrapped_attrib(ec_attribute_t *wrapped_attrib, uint8_t *frame, size_t frame_len, uint8_t *frame_attribs, bool uses_aad, uint8_t *key);
+	static std::pair<uint8_t*, uint16_t> unwrap_wrapped_attrib(ec_attribute_t wrapped_attrib, uint8_t *frame, size_t frame_len, uint8_t *frame_attribs, bool uses_aad, uint8_t *key);
```
Same `const` ref comment here

bcarlson-dev commented on src/em/prov/easyconnect/ec_util.cpp:
```diff
@@ -24,25 +24,35 @@ void ec_util::init_frame(ec_frame_t *frame)
 }
 
-ec_attribute_t *ec_util::get_attrib(uint8_t *buff, size_t len, ec_attrib_id_t id)
+std::optional<ec_attribute_t> ec_util::get_attrib(uint8_t *buff, size_t len, ec_attrib_id_t id)
```
Based on the prior comment, saying that to pass in `ec_attribute_t` as a `const` ref, it might make it easier to make the return type here be a `const std::optional<ec_attribute_t>`. It would also make it more clear since none of the `ec_attribute_t`s parameters should be modified since it is just a helper struct. The actual buffer data is in `ec_net_attribute_t` and (therefore) `data` (whose pointer also can't be modified with a `const`).

apahl-cl replied:
I tested returning `const std::optional<ec_attribute_t>` from `get_attrib`. It still allows modifications - I think the intuition is that the subsequent assignment, `auto foo = get_attrib(...)`, counts as a copy and discards the const qualifier (not 100% sure though).

Protecting the struct from accidental modifications seems a bit complicated - I think what we really want is to return `std::optional<const ec_attribute_t>`. Or, for completeness, we can return `const std::optional<const ec_attribute_t>` AND also use `const auto foo = ...` on the caller's side.

apahl-cl replied:
Okay, I've changed the return value to `const std::optional<const ec_attribute_t>` and I'm using `const auto foo = ...` on the caller side. We can adjust further if needed.

bcarlson-dev commented on tests/test_ec_attrib.cpp:
```diff
@@ -105,30 +118,33 @@ TEST_F(ECUtilAttributeTest, GetAttribute) {
     EXPECT_EQ(attr2->length, sizeof(uint16_t));
+    EXPECT_EQ(attr2->original, reinterpret_cast<ec_net_attribute_t*>(test_attrib_2));
 
     uint16_t data = 0;
     memcpy(&data, attr2->data, sizeof(uint16_t));
```
We can use one of your new `deref` calls if you want here unless the goal is to keep those functions out intentionally

apahl-cl replied:
I don't think that will work here actually - the new `deref` calls use `ntohs`, but the input is network byte ordered and we want the output to keep that.

bcarlson-dev commented on tests/test_ec_attrib.cpp:
```diff
@@ -45,22 +45,35 @@ class ECUtilAttributeTest : public ::testing::Test {
+    const uint16_t test_val_2 = 0x1234;
+    const std::string test_val_3 = "test_string";
+
+    uint8_t *test_attrib_1;
+    uint8_t *test_attrib_2;
+    uint8_t *test_attrib_3;
```
Might be worth while to just set these to `NULL` in the `TearDown` function

bcarlson-dev reviewed (changes requested):
Honestly it looks pretty near perfect. Since this is a pretty crucial piece of code I want to be really through so I appreciate your patience.

apahl-cl commented:
Implemented most of the changes from Ben's latest code review. Modified `TEST_F(ECUtilAttributeTest, NullBuffer)` (`test_ec_attrib.cpp:261`) because an attribute was being reassigned in that test.

bcarlson-dev commented on src/em/prov/easyconnect/ec_util.cpp:
```diff
@@ -24,25 +24,35 @@ void ec_util::init_frame(ec_frame_t *frame)
 }
 
-ec_attribute_t *ec_util::get_attrib(uint8_t *buff, size_t len, ec_attrib_id_t id)
+const std::optional<const ec_attribute_t> ec_util::get_attrib(uint8_t *buff, size_t len, ec_attrib_id_t id)
```
So I think you were right and we don't need the `const` for the optional return type. Removing it will also remove the need for the `const auto` as well. If I am not mistaken, keeping the `const ec_attribute_t` should make it so that when you do the `*some_attrib` later on you won't have to do a cast so we should be good if you remove this and switch back from the `const auto` to the `auto`?
```suggestion
std::optional<const ec_attribute_t> ec_util::get_attrib(uint8_t *buff, size_t len, ec_attrib_id_t id)
```

tuckerpo-cl commented:
Thanks for plugging all the leaks I created 🙏  C moment!

apahl-cl commented:
> Thanks for plugging all the leaks I created 🙏 C moment!

Glad i could help a bit, and thanks to Ben too for plugging all the leaks I created to get this far!
